workflow/src/main/java/bio/terra/tanagra/workflow/GraphUtils.java

```diff
@@ -0,0 +1,131 @@
+   *
+   * @param edges A directed acyclic graph where every KV defines a directed edge with a {Head,
+   *     Tail} vertex pair.
+   * @param maxPathLength The maximum path length to try to find in the graph. If there is a path in
+   *     the graph longer than this, not all paths will be found. This is used to bound execution.
+   * @param <T> The type of the vertex in the graph.
+   * @return A PCollection of all reachable vertex pairs, where there may be duplicate pairs.
+   */
+  public static <T> PCollection<KV<T, T>> transitiveClosure(
+      PCollection<KV<T, T>> edges, int maxPathLength) {
+    // allPaths[n] is the set of vertices {V0, V1}, where V0 -> V1 is reachable in N or less edges.
+    // That is, there is a path with N or less edges from V0 to V1.
+    Map<Integer, PCollection<KV<T, T>>> allPaths = new HashMap<>();
+    // exactPaths[n] is the set of vertices where there exists a path from V0 -> V1 in exactly N
+    // edges.
+    Map<Integer, PCollection<KV<T, T>>> exactPaths = new HashMap<>();
+
+    // The input edges are both allPaths[1] and exactPaths[1].
+    allPaths.put(1, edges);
+    exactPaths.put(1, edges);
+
+    // Iteratively build up allPaths in O(log_2(maxPathLength)) joins. We want to avoid doing many
+    // joins or creating many duplicate paths that will increase the runtime.
+    // The simplest algorithm would be to compute all paths incrementally one edge length at a time,
+    // taking O(maxPathLength) joins:
+    // allPaths[x+1] = allPaths[x] + exactPaths[x] * allPaths[1],
+    // where '+' is appending lists, and '*' is concatenating paths together, joining on vertexes
+    // e.g. [{A, B}] * [{B, C}, {B, D}] yields [{A, C}, {A,D}].
+    //
+    // Given that exactPaths[x + y] = exactPaths[x] * exactPaths[y], and
+    // and that for x >= y, allPaths[x] = allPaths[y - 1] + exactPaths[y] * allPaths[x - y], we have
+    // allPaths[2n - 1] = allPaths[n - 1] + exactPaths[n] + exactPaths[n] * allPaths[n - 1].
+    //
+    // With this, we can increase n exponentially by 2, allowing for O(log_2(maxPathLengths) joins,
+    // (n=2) allPaths[3] = allPaths[1] + exactPaths[2] + exactPaths[2] * allPaths[1]
+    // (n=4) allPaths[7] = allPaths[3] + exactPaths[4] + exactPaths[4] * allPaths[3]
+    // etc. where exactPaths[n] = exactPaths[n/2] * exactPaths[n/2].
+    int n = 2;
+    while (true) {
+      PCollection<KV<T, T>> nExactPaths =
+          concatenate(exactPaths.get(n / 2), exactPaths.get(n / 2), "exactPaths N" + n);
+      exactPaths.put(n, nExactPaths);
+
+      PCollection<KV<T, T>> nMinus1AllPaths = allPaths.get(n - 1);
+
+      PCollection<KV<T, T>> newAllPaths =
+          PCollectionList.of(nMinus1AllPaths)
+              .and(nExactPaths)
+              .and(concatenate(nExactPaths, nMinus1AllPaths, "allPaths temp N" + n))
+              .apply("flatten N" + n, Flatten.pCollections());
+      int newLongestPathLength = 2 * n - 1;
+      if (newLongestPathLength >= maxPathLength) {
```

Comment:
nit: maybe rename `maxPathLength` to make it clearer that this is to bound computation, and is not an exact cutoff to the answer e.g. `maxPathLengthHint`?

Reply:
thanks, done.